// v-network-graph/lib/force-layout.mjs
import { watch as yt, toRef as gt } from "vue";
function et(t, n) {
  var e, r = 1;
  t == null && (t = 0), n == null && (n = 0);
  function i() {
    var o, l = e.length, a, h = 0, f = 0;
    for (o = 0; o < l; ++o)
      a = e[o], h += a.x, f += a.y;
    for (h = (h / l - t) * r, f = (f / l - n) * r, o = 0; o < l; ++o)
      a = e[o], a.x -= h, a.y -= f;
  }
  return i.initialize = function(o) {
    e = o;
  }, i.x = function(o) {
    return arguments.length ? (t = +o, i) : t;
  }, i.y = function(o) {
    return arguments.length ? (n = +o, i) : n;
  }, i.strength = function(o) {
    return arguments.length ? (r = +o, i) : r;
  }, i;
}
function vt(t) {
  const n = +this._x.call(null, t), e = +this._y.call(null, t);
  return rt(this.cover(n, e), n, e, t);
}
function rt(t, n, e, r) {
  if (isNaN(n) || isNaN(e))
    return t;
  var i, o = t._root, l = { data: r }, a = t._x0, h = t._y0, f = t._x1, v = t._y1, w, x, g, _, s, u, c, y;
  if (!o)
    return t._root = l, t;
  for (; o.length; )
    if ((s = n >= (w = (a + f) / 2)) ? a = w : f = w, (u = e >= (x = (h + v) / 2)) ? h = x : v = x, i = o, !(o = o[c = u << 1 | s]))
      return i[c] = l, t;
  if (g = +t._x.call(null, o.data), _ = +t._y.call(null, o.data), n === g && e === _)
    return l.next = o, i ? i[c] = l : t._root = l, t;
  do
    i = i ? i[c] = new Array(4) : t._root = new Array(4), (s = n >= (w = (a + f) / 2)) ? a = w : f = w, (u = e >= (x = (h + v) / 2)) ? h = x : v = x;
  while ((c = u << 1 | s) === (y = (_ >= x) << 1 | g >= w));
  return i[y] = o, i[c] = l, t;
}
function pt(t) {
  var n, e, r = t.length, i, o, l = new Array(r), a = new Array(r), h = 1 / 0, f = 1 / 0, v = -1 / 0, w = -1 / 0;
  for (e = 0; e < r; ++e)
    isNaN(i = +this._x.call(null, n = t[e])) || isNaN(o = +this._y.call(null, n)) || (l[e] = i, a[e] = o, i < h && (h = i), i > v && (v = i), o < f && (f = o), o > w && (w = o));
  if (h > v || f > w)
    return this;
  for (this.cover(h, f).cover(v, w), e = 0; e < r; ++e)
    rt(this, l[e], a[e], t[e]);
  return this;
}
function xt(t, n) {
  if (isNaN(t = +t) || isNaN(n = +n))
    return this;
  var e = this._x0, r = this._y0, i = this._x1, o = this._y1;
  if (isNaN(e))
    i = (e = Math.floor(t)) + 1, o = (r = Math.floor(n)) + 1;
  else {
    for (var l = i - e || 1, a = this._root, h, f; e > t || t >= i || r > n || n >= o; )
      switch (f = (n < r) << 1 | t < e, h = new Array(4), h[f] = a, a = h, l *= 2, f) {
        case 0:
          i = e + l, o = r + l;
          break;
        case 1:
          e = i - l, o = r + l;
          break;
        case 2:
          i = e + l, r = o - l;
          break;
        case 3:
          e = i - l, r = o - l;
          break;
      }
    this._root && this._root.length && (this._root = a);
  }
  return this._x0 = e, this._y0 = r, this._x1 = i, this._y1 = o, this;
}
function _t() {
  var t = [];
  return this.visit(function(n) {
    if (!n.length)
      do
        t.push(n.data);
      while (n = n.next);
  }), t;
}
function wt(t) {
  return arguments.length ? this.cover(+t[0][0], +t[0][1]).cover(+t[1][0], +t[1][1]) : isNaN(this._x0) ? void 0 : [[this._x0, this._y0], [this._x1, this._y1]];
}
function O(t, n, e, r, i) {
  this.node = t, this.x0 = n, this.y0 = e, this.x1 = r, this.y1 = i;
}
function dt(t, n, e) {
  var r, i = this._x0, o = this._y0, l, a, h, f, v = this._x1, w = this._y1, x = [], g = this._root, _, s;
  for (g && x.push(new O(g, i, o, v, w)), e == null ? e = 1 / 0 : (i = t - e, o = n - e, v = t + e, w = n + e, e *= e); _ = x.pop(); )
    if (!(!(g = _.node) || (l = _.x0) > v || (a = _.y0) > w || (h = _.x1) < i || (f = _.y1) < o))
      if (g.length) {
        var u = (l + h) / 2, c = (a + f) / 2;
        x.push(
          new O(g[3], u, c, h, f),
          new O(g[2], l, c, u, f),
          new O(g[1], u, a, h, c),
          new O(g[0], l, a, u, c)
        ), (s = (n >= c) << 1 | t >= u) && (_ = x[x.length - 1], x[x.length - 1] = x[x.length - 1 - s], x[x.length - 1 - s] = _);
      } else {
        var y = t - +this._x.call(null, g.data), d = n - +this._y.call(null, g.data), p = y * y + d * d;
        if (p < e) {
          var m = Math.sqrt(e = p);
          i = t - m, o = n - m, v = t + m, w = n + m, r = g.data;
        }
      }
  return r;
}
function mt(t) {
  if (isNaN(v = +this._x.call(null, t)) || isNaN(w = +this._y.call(null, t)))
    return this;
  var n, e = this._root, r, i, o, l = this._x0, a = this._y0, h = this._x1, f = this._y1, v, w, x, g, _, s, u, c;
  if (!e)
    return this;
  if (e.length)
    for (; ; ) {
      if ((_ = v >= (x = (l + h) / 2)) ? l = x : h = x, (s = w >= (g = (a + f) / 2)) ? a = g : f = g, n = e, !(e = e[u = s << 1 | _]))
        return this;
      if (!e.length)
        break;
      (n[u + 1 & 3] || n[u + 2 & 3] || n[u + 3 & 3]) && (r = n, c = u);
    }
  for (; e.data !== t; )
    if (i = e, !(e = e.next))
      return this;
  return (o = e.next) && delete e.next, i ? (o ? i.next = o : delete i.next, this) : n ? (o ? n[u] = o : delete n[u], (e = n[0] || n[1] || n[2] || n[3]) && e === (n[3] || n[2] || n[1] || n[0]) && !e.length && (r ? r[c] = e : this._root = e), this) : (this._root = o, this);
}
function Nt(t) {
  for (var n = 0, e = t.length; n < e; ++n)
    this.remove(t[n]);
  return this;
}
function At() {
  return this._root;
}
function Mt() {
  var t = 0;
  return this.visit(function(n) {
    if (!n.length)
      do
        ++t;
      while (n = n.next);
  }), t;
}
function bt(t) {
  var n = [], e, r = this._root, i, o, l, a, h;
  for (r && n.push(new O(r, this._x0, this._y0, this._x1, this._y1)); e = n.pop(); )
    if (!t(r = e.node, o = e.x0, l = e.y0, a = e.x1, h = e.y1) && r.length) {
      var f = (o + a) / 2, v = (l + h) / 2;
      (i = r[3]) && n.push(new O(i, f, v, a, h)), (i = r[2]) && n.push(new O(i, o, v, f, h)), (i = r[1]) && n.push(new O(i, f, l, a, v)), (i = r[0]) && n.push(new O(i, o, l, f, v));
    }
  return this;
}
function Ot(t) {
  var n = [], e = [], r;
  for (this._root && n.push(new O(this._root, this._x0, this._y0, this._x1, this._y1)); r = n.pop(); ) {
    var i = r.node;
    if (i.length) {
      var o, l = r.x0, a = r.y0, h = r.x1, f = r.y1, v = (l + h) / 2, w = (a + f) / 2;
      (o = i[0]) && n.push(new O(o, l, a, v, w)), (o = i[1]) && n.push(new O(o, v, a, h, w)), (o = i[2]) && n.push(new O(o, l, w, v, f)), (o = i[3]) && n.push(new O(o, v, w, h, f));
    }
    e.push(r);
  }
  for (; r = e.pop(); )
    t(r.node, r.x0, r.y0, r.x1, r.y1);
  return this;
}
function jt(t) {
  return t[0];
}
function Pt(t) {
  return arguments.length ? (this._x = t, this) : this._x;
}
function zt(t) {
  return t[1];
}
function Lt(t) {
  return arguments.length ? (this._y = t, this) : this._y;
}
function Y(t, n, e) {
  var r = new H(n == null ? jt : n, e == null ? zt : e, NaN, NaN, NaN, NaN);
  return t == null ? r : r.addAll(t);
}
function H(t, n, e, r, i, o) {
  this._x = t, this._y = n, this._x0 = e, this._y0 = r, this._x1 = i, this._y1 = o, this._root = void 0;
}
function Z(t) {
  for (var n = { data: t.data }, e = n; t = t.next; )
    e = e.next = { data: t.data };
  return n;
}
var j = Y.prototype = H.prototype;
j.copy = function() {
  var t = new H(this._x, this._y, this._x0, this._y0, this._x1, this._y1), n = this._root, e, r;
  if (!n)
    return t;
  if (!n.length)
    return t._root = Z(n), t;
  for (e = [{ source: n, target: t._root = new Array(4) }]; n = e.pop(); )
    for (var i = 0; i < 4; ++i)
      (r = n.source[i]) && (r.length ? e.push({ source: r, target: n.target[i] = new Array(4) }) : n.target[i] = Z(r));
  return t;
};
j.add = vt;
j.addAll = pt;
j.cover = xt;
j.data = _t;
j.extent = wt;
j.find = dt;
j.remove = mt;
j.removeAll = Nt;
j.root = At;
j.size = Mt;
j.visit = bt;
j.visitAfter = Ot;
j.x = Pt;
j.y = Lt;
function N(t) {
  return function() {
    return t;
  };
}
function I(t) {
  return (t() - 0.5) * 1e-6;
}
function Et(t) {
  return t.x + t.vx;
}
function Dt(t) {
  return t.y + t.vy;
}
function it(t) {
  var n, e, r, i = 1, o = 1;
  typeof t != "function" && (t = N(t == null ? 1 : +t));
  function l() {
    for (var f, v = n.length, w, x, g, _, s, u, c = 0; c < o; ++c)
      for (w = Y(n, Et, Dt).visitAfter(a), f = 0; f < v; ++f)
        x = n[f], s = e[x.index], u = s * s, g = x.x + x.vx, _ = x.y + x.vy, w.visit(y);
    function y(d, p, m, b, P) {
      var A = d.data, z = d.r, M = s + z;
      if (A) {
        if (A.index > x.index) {
          var E = g - A.x - A.vx, D = _ - A.y - A.vy, L = E * E + D * D;
          L < M * M && (E === 0 && (E = I(r), L += E * E), D === 0 && (D = I(r), L += D * D), L = (M - (L = Math.sqrt(L))) / L * i, x.vx += (E *= L) * (M = (z *= z) / (u + z)), x.vy += (D *= L) * M, A.vx -= E * (M = 1 - M), A.vy -= D * M);
        }
        return;
      }
      return p > g + M || b < g - M || m > _ + M || P < _ - M;
    }
  }
  function a(f) {
    if (f.data)
      return f.r = e[f.data.index];
    for (var v = f.r = 0; v < 4; ++v)
      f[v] && f[v].r > f.r && (f.r = f[v].r);
  }
  function h() {
    if (!!n) {
      var f, v = n.length, w;
      for (e = new Array(v), f = 0; f < v; ++f)
        w = n[f], e[w.index] = +t(w, f, n);
    }
  }
  return l.initialize = function(f, v) {
    n = f, r = v, h();
  }, l.iterations = function(f) {
    return arguments.length ? (o = +f, l) : o;
  }, l.strength = function(f) {
    return arguments.length ? (i = +f, l) : i;
  }, l.radius = function(f) {
    return arguments.length ? (t = typeof f == "function" ? f : N(+f), h(), l) : t;
  }, l;
}
function It(t) {
  return t.index;
}
function G(t, n) {
  var e = t.get(n);
  if (!e)
    throw new Error("node not found: " + n);
  return e;
}
function ot(t) {
  var n = It, e = w, r, i = N(30), o, l, a, h, f, v = 1;
  t == null && (t = []);
  function w(u) {
    return 1 / Math.min(a[u.source.index], a[u.target.index]);
  }
  function x(u) {
    for (var c = 0, y = t.length; c < v; ++c)
      for (var d = 0, p, m, b, P, A, z, M; d < y; ++d)
        p = t[d], m = p.source, b = p.target, P = b.x + b.vx - m.x - m.vx || I(f), A = b.y + b.vy - m.y - m.vy || I(f), z = Math.sqrt(P * P + A * A), z = (z - o[d]) / z * u * r[d], P *= z, A *= z, b.vx -= P * (M = h[d]), b.vy -= A * M, m.vx += P * (M = 1 - M), m.vy += A * M;
  }
  function g() {
    if (!!l) {
      var u, c = l.length, y = t.length, d = new Map(l.map((m, b) => [n(m, b, l), m])), p;
      for (u = 0, a = new Array(c); u < y; ++u)
        p = t[u], p.index = u, typeof p.source != "object" && (p.source = G(d, p.source)), typeof p.target != "object" && (p.target = G(d, p.target)), a[p.source.index] = (a[p.source.index] || 0) + 1, a[p.target.index] = (a[p.target.index] || 0) + 1;
      for (u = 0, h = new Array(y); u < y; ++u)
        p = t[u], h[u] = a[p.source.index] / (a[p.source.index] + a[p.target.index]);
      r = new Array(y), _(), o = new Array(y), s();
    }
  }
  function _() {
    if (!!l)
      for (var u = 0, c = t.length; u < c; ++u)
        r[u] = +e(t[u], u, t);
  }
  function s() {
    if (!!l)
      for (var u = 0, c = t.length; u < c; ++u)
        o[u] = +i(t[u], u, t);
  }
  return x.initialize = function(u, c) {
    l = u, f = c, g();
  }, x.links = function(u) {
    return arguments.length ? (t = u, g(), x) : t;
  }, x.id = function(u) {
    return arguments.length ? (n = u, x) : n;
  }, x.iterations = function(u) {
    return arguments.length ? (v = +u, x) : v;
  }, x.strength = function(u) {
    return arguments.length ? (e = typeof u == "function" ? u : N(+u), _(), x) : e;
  }, x.distance = function(u) {
    return arguments.length ? (i = typeof u == "function" ? u : N(+u), s(), x) : i;
  }, x;
}
var St = { value: () => {
} };
function at() {
  for (var t = 0, n = arguments.length, e = {}, r; t < n; ++t) {
    if (!(r = arguments[t] + "") || r in e || /[\s.]/.test(r))
      throw new Error("illegal type: " + r);
    e[r] = [];
  }
  return new C(e);
}
function C(t) {
  this._ = t;
}
function Tt(t, n) {
  return t.trim().split(/^|\s+/).map(function(e) {
    var r = "", i = e.indexOf(".");
    if (i >= 0 && (r = e.slice(i + 1), e = e.slice(0, i)), e && !n.hasOwnProperty(e))
      throw new Error("unknown type: " + e);
    return { type: e, name: r };
  });
}
C.prototype = at.prototype = {
  constructor: C,
  on: function(t, n) {
    var e = this._, r = Tt(t + "", e), i, o = -1, l = r.length;
    if (arguments.length < 2) {
      for (; ++o < l; )
        if ((i = (t = r[o]).type) && (i = kt(e[i], t.name)))
          return i;
      return;
    }
    if (n != null && typeof n != "function")
      throw new Error("invalid callback: " + n);
    for (; ++o < l; )
      if (i = (t = r[o]).type)
        e[i] = J(e[i], t.name, n);
      else if (n == null)
        for (i in e)
          e[i] = J(e[i], t.name, null);
    return this;
  },
  copy: function() {
    var t = {}, n = this._;
    for (var e in n)
      t[e] = n[e].slice();
    return new C(t);
  },
  call: function(t, n) {
    if ((i = arguments.length - 2) > 0)
      for (var e = new Array(i), r = 0, i, o; r < i; ++r)
        e[r] = arguments[r + 2];
    if (!this._.hasOwnProperty(t))
      throw new Error("unknown type: " + t);
    for (o = this._[t], r = 0, i = o.length; r < i; ++r)
      o[r].value.apply(n, e);
  },
  apply: function(t, n, e) {
    if (!this._.hasOwnProperty(t))
      throw new Error("unknown type: " + t);
    for (var r = this._[t], i = 0, o = r.length; i < o; ++i)
      r[i].value.apply(n, e);
  }
};
function kt(t, n) {
  for (var e = 0, r = t.length, i; e < r; ++e)
    if ((i = t[e]).name === n)
      return i.value;
}
function J(t, n, e) {
  for (var r = 0, i = t.length; r < i; ++r)
    if (t[r].name === n) {
      t[r] = St, t = t.slice(0, r).concat(t.slice(r + 1));
      break;
    }
  return e != null && t.push({ name: n, value: e }), t;
}
var T = 0, F = 0, k = 0, ft = 1e3, R, $, K = 0, S = 0, Q = 0, B = typeof performance == "object" && performance.now ? performance : Date, ut = typeof window == "object" && window.requestAnimationFrame ? window.requestAnimationFrame.bind(window) : function(t) {
  setTimeout(t, 17);
};
function st() {
  return S || (ut(Ft), S = B.now() + Q);
}
function Ft() {
  S = 0;
}
function W() {
  this._call = this._time = this._next = null;
}
W.prototype = lt.prototype = {
  constructor: W,
  restart: function(t, n, e) {
    if (typeof t != "function")
      throw new TypeError("callback is not a function");
    e = (e == null ? st() : +e) + (n == null ? 0 : +n), !this._next && $ !== this && ($ ? $._next = this : R = this, $ = this), this._call = t, this._time = e, X();
  },
  stop: function() {
    this._call && (this._call = null, this._time = 1 / 0, X());
  }
};
function lt(t, n, e) {
  var r = new W();
  return r.restart(t, n, e), r;
}
function $t() {
  st(), ++T;
  for (var t = R, n; t; )
    (n = S - t._time) >= 0 && t._call.call(void 0, n), t = t._next;
  --T;
}
function U() {
  S = (K = B.now()) + Q, T = F = 0;
  try {
    $t();
  } finally {
    T = 0, Ct(), S = 0;
  }
}
function Bt() {
  var t = B.now(), n = t - K;
  n > ft && (Q -= n, K = t);
}
function Ct() {
  for (var t, n = R, e, r = 1 / 0; n; )
    n._call ? (r > n._time && (r = n._time), t = n, n = n._next) : (e = n._next, n._next = null, n = t ? t._next = e : R = e);
  $ = t, X(r);
}
function X(t) {
  if (!T) {
    F && (F = clearTimeout(F));
    var n = t - S;
    n > 24 ? (t < 1 / 0 && (F = setTimeout(U, t - B.now() - Q)), k && (k = clearInterval(k))) : (k || (K = B.now(), k = setInterval(Bt, ft)), T = 1, ut(U));
  }
}
const Rt = 1664525, Kt = 1013904223, q = 4294967296;
function Qt() {
  let t = 1;
  return () => (t = (Rt * t + Kt) % q) / q;
}
function Vt(t) {
  return t.x;
}
function Wt(t) {
  return t.y;
}
var Xt = 10, Yt = Math.PI * (3 - Math.sqrt(5));
function ct(t) {
  var n, e = 1, r = 1e-3, i = 1 - Math.pow(r, 1 / 300), o = 0, l = 0.6, a = /* @__PURE__ */ new Map(), h = lt(w), f = at("tick", "end"), v = Qt();
  t == null && (t = []);
  function w() {
    x(), f.call("tick", n), e < r && (h.stop(), f.call("end", n));
  }
  function x(s) {
    var u, c = t.length, y;
    s === void 0 && (s = 1);
    for (var d = 0; d < s; ++d)
      for (e += (o - e) * i, a.forEach(function(p) {
        p(e);
      }), u = 0; u < c; ++u)
        y = t[u], y.fx == null ? y.x += y.vx *= l : (y.x = y.fx, y.vx = 0), y.fy == null ? y.y += y.vy *= l : (y.y = y.fy, y.vy = 0);
    return n;
  }
  function g() {
    for (var s = 0, u = t.length, c; s < u; ++s) {
      if (c = t[s], c.index = s, c.fx != null && (c.x = c.fx), c.fy != null && (c.y = c.fy), isNaN(c.x) || isNaN(c.y)) {
        var y = Xt * Math.sqrt(0.5 + s), d = s * Yt;
        c.x = y * Math.cos(d), c.y = y * Math.sin(d);
      }
      (isNaN(c.vx) || isNaN(c.vy)) && (c.vx = c.vy = 0);
    }
  }
  function _(s) {
    return s.initialize && s.initialize(t, v), s;
  }
  return g(), n = {
    tick: x,
    restart: function() {
      return h.restart(w), n;
    },
    stop: function() {
      return h.stop(), n;
    },
    nodes: function(s) {
      return arguments.length ? (t = s, g(), a.forEach(_), n) : t;
    },
    alpha: function(s) {
      return arguments.length ? (e = +s, n) : e;
    },
    alphaMin: function(s) {
      return arguments.length ? (r = +s, n) : r;
    },
    alphaDecay: function(s) {
      return arguments.length ? (i = +s, n) : +i;
    },
    alphaTarget: function(s) {
      return arguments.length ? (o = +s, n) : o;
    },
    velocityDecay: function(s) {
      return arguments.length ? (l = 1 - s, n) : 1 - l;
    },
    randomSource: function(s) {
      return arguments.length ? (v = s, a.forEach(_), n) : v;
    },
    force: function(s, u) {
      return arguments.length > 1 ? (u == null ? a.delete(s) : a.set(s, _(u)), n) : a.get(s);
    },
    find: function(s, u, c) {
      var y = 0, d = t.length, p, m, b, P, A;
      for (c == null ? c = 1 / 0 : c *= c, y = 0; y < d; ++y)
        P = t[y], p = s - P.x, m = u - P.y, b = p * p + m * m, b < c && (A = P, c = b);
      return A;
    },
    on: function(s, u) {
      return arguments.length > 1 ? (f.on(s, u), n) : f.on(s);
    }
  };
}
function ht() {
  var t, n, e, r, i = N(-30), o, l = 1, a = 1 / 0, h = 0.81;
  function f(g) {
    var _, s = t.length, u = Y(t, Vt, Wt).visitAfter(w);
    for (r = g, _ = 0; _ < s; ++_)
      n = t[_], u.visit(x);
  }
  function v() {
    if (!!t) {
      var g, _ = t.length, s;
      for (o = new Array(_), g = 0; g < _; ++g)
        s = t[g], o[s.index] = +i(s, g, t);
    }
  }
  function w(g) {
    var _ = 0, s, u, c = 0, y, d, p;
    if (g.length) {
      for (y = d = p = 0; p < 4; ++p)
        (s = g[p]) && (u = Math.abs(s.value)) && (_ += s.value, c += u, y += u * s.x, d += u * s.y);
      g.x = y / c, g.y = d / c;
    } else {
      s = g, s.x = s.data.x, s.y = s.data.y;
      do
        _ += o[s.data.index];
      while (s = s.next);
    }
    g.value = _;
  }
  function x(g, _, s, u) {
    if (!g.value)
      return !0;
    var c = g.x - n.x, y = g.y - n.y, d = u - _, p = c * c + y * y;
    if (d * d / h < p)
      return p < a && (c === 0 && (c = I(e), p += c * c), y === 0 && (y = I(e), p += y * y), p < l && (p = Math.sqrt(l * p)), n.vx += c * g.value * r / p, n.vy += y * g.value * r / p), !0;
    if (g.length || p >= a)
      return;
    (g.data !== n || g.next) && (c === 0 && (c = I(e), p += c * c), y === 0 && (y = I(e), p += y * y), p < l && (p = Math.sqrt(l * p)));
    do
      g.data !== n && (d = o[g.data.index] * r / p, n.vx += c * d, n.vy += y * d);
    while (g = g.next);
  }
  return f.initialize = function(g, _) {
    t = g, e = _, v();
  }, f.strength = function(g) {
    return arguments.length ? (i = typeof g == "function" ? g : N(+g), v(), f) : i;
  }, f.distanceMin = function(g) {
    return arguments.length ? (l = g * g, f) : Math.sqrt(l);
  }, f.distanceMax = function(g) {
    return arguments.length ? (a = g * g, f) : Math.sqrt(a);
  }, f.theta = function(g) {
    return arguments.length ? (h = g * g, f) : Math.sqrt(h);
  }, f;
}
function Ht(t, n, e) {
  var r, i = N(0.1), o, l;
  typeof t != "function" && (t = N(+t)), n == null && (n = 0), e == null && (e = 0);
  function a(f) {
    for (var v = 0, w = r.length; v < w; ++v) {
      var x = r[v], g = x.x - n || 1e-6, _ = x.y - e || 1e-6, s = Math.sqrt(g * g + _ * _), u = (l[v] - s) * o[v] * f / s;
      x.vx += g * u, x.vy += _ * u;
    }
  }
  function h() {
    if (!!r) {
      var f, v = r.length;
      for (o = new Array(v), l = new Array(v), f = 0; f < v; ++f)
        l[f] = +t(r[f], f, r), o[f] = isNaN(l[f]) ? 0 : +i(r[f], f, r);
    }
  }
  return a.initialize = function(f) {
    r = f, h();
  }, a.strength = function(f) {
    return arguments.length ? (i = typeof f == "function" ? f : N(+f), h(), a) : i;
  }, a.radius = function(f) {
    return arguments.length ? (t = typeof f == "function" ? f : N(+f), h(), a) : t;
  }, a.x = function(f) {
    return arguments.length ? (n = +f, a) : n;
  }, a.y = function(f) {
    return arguments.length ? (e = +f, a) : e;
  }, a;
}
function Zt(t) {
  var n = N(0.1), e, r, i;
  typeof t != "function" && (t = N(t == null ? 0 : +t));
  function o(a) {
    for (var h = 0, f = e.length, v; h < f; ++h)
      v = e[h], v.vx += (i[h] - v.x) * r[h] * a;
  }
  function l() {
    if (!!e) {
      var a, h = e.length;
      for (r = new Array(h), i = new Array(h), a = 0; a < h; ++a)
        r[a] = isNaN(i[a] = +t(e[a], a, e)) ? 0 : +n(e[a], a, e);
    }
  }
  return o.initialize = function(a) {
    e = a, l();
  }, o.strength = function(a) {
    return arguments.length ? (n = typeof a == "function" ? a : N(+a), l(), o) : n;
  }, o.x = function(a) {
    return arguments.length ? (t = typeof a == "function" ? a : N(+a), l(), o) : t;
  }, o;
}
function Gt(t) {
  var n = N(0.1), e, r, i;
  typeof t != "function" && (t = N(t == null ? 0 : +t));
  function o(a) {
    for (var h = 0, f = e.length, v; h < f; ++h)
      v = e[h], v.vy += (i[h] - v.y) * r[h] * a;
  }
  function l() {
    if (!!e) {
      var a, h = e.length;
      for (r = new Array(h), i = new Array(h), a = 0; a < h; ++a)
        r[a] = isNaN(i[a] = +t(e[a], a, e)) ? 0 : +n(e[a], a, e);
    }
  }
  return o.initialize = function(a) {
    e = a, l();
  }, o.strength = function(a) {
    return arguments.length ? (n = typeof a == "function" ? a : N(+a), l(), o) : n;
  }, o.y = function(a) {
    return arguments.length ? (t = typeof a == "function" ? a : N(+a), l(), o) : t;
  }, o;
}
const Jt = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  forceCenter: et,
  forceCollide: it,
  forceLink: ot,
  forceManyBody: ht,
  forceRadial: Ht,
  forceSimulation: ct,
  forceX: Zt,
  forceY: Gt
}, Symbol.toStringTag, { value: "Module" }));
var Ut = Object.defineProperty, qt = Object.defineProperties, tn = Object.getOwnPropertyDescriptors, tt = Object.getOwnPropertySymbols, nn = Object.prototype.hasOwnProperty, en = Object.prototype.propertyIsEnumerable, nt = (t, n, e) => n in t ? Ut(t, n, { enumerable: !0, configurable: !0, writable: !0, value: e }) : t[n] = e, V = (t, n) => {
  for (var e in n || (n = {}))
    nn.call(n, e) && nt(t, e, n[e]);
  if (tt)
    for (var e of tt(n))
      en.call(n, e) && nt(t, e, n[e]);
  return t;
}, rn = (t, n) => qt(t, tn(n));
class an {
  constructor(n = {}) {
    this.options = n;
  }
  activate(n) {
    const { nodePositions: e, nodes: r, edges: i, emitter: o, svgPanZoom: l } = n;
    let { nodeLayouts: a, nodeLayoutMap: h } = this.buildNodeLayouts(r.value, e, {
      x: 0,
      y: 0
    });
    const f = this.createSimulation(
      a,
      this.forceLayoutEdges(i.value, r.value)
    );
    f.on("tick", () => {
      var s, u, c;
      for (const y of a) {
        const d = (s = e.value) == null ? void 0 : s[y.id];
        d && (d.x = (u = y.x) != null ? u : 0, d.y = (c = y.y) != null ? c : 0);
      }
    });
    const v = () => {
      f.alpha(0.1).restart();
    }, w = (s) => {
      for (const [u, c] of Object.entries(s)) {
        const y = h[u];
        y.fx = c.x, y.fy = c.y;
      }
      v();
    }, x = (s) => {
      var u;
      for (const [c, y] of Object.entries(s)) {
        const d = this.getNodeLayout(e, c), p = (u = h == null ? void 0 : h[c]) != null ? u : { x: 0, y: 0 };
        d.value.fixed || this.options.positionFixedByDrag ? (p.fx = y.x, p.fy = y.y, d.value.fixed = !0) : (p.x = y.x, p.y = y.y, delete p.fx, delete p.fy);
      }
      v();
    }, g = ({ node: s, event: u }) => {
      if (this.options.positionFixedByClickWithAltKey && u.altKey) {
        const c = this.getNodeLayout(e, s);
        let y = h == null ? void 0 : h[s];
        y || (y = { id: s, x: 0, y: 0 }, h[s] = y), c.value.fixed ? (delete c.value.fixed, y.x = y.fx || y.x, y.y = y.fy || y.y, delete y.fx, delete y.fy) : (c.value.fixed = !0, y.fx = y.x, y.fy = y.y), v();
      }
    }, _ = yt(
      () => [
        Object.keys(r.value),
        Object.values(i.value).map((s) => `${s.source}=${s.target}`)
      ],
      () => {
        const s = l.getViewArea();
        ({ nodeLayouts: a, nodeLayoutMap: h } = this.buildNodeLayouts(
          r.value,
          e,
          s.center
        )), f.nodes(a);
        const u = f.force("edge");
        u && u.links(this.forceLayoutEdges(i.value, r.value)), v();
      }
    );
    o.on("node:dragstart", w), o.on("node:pointermove", w), o.on("node:dragend", x), o.on("node:click", g), this.onDeactivate = () => {
      f.stop(), _(), o.off("node:dragstart", w), o.off("node:pointermove", w), o.off("node:dragend", x), o.off("node:click", g);
    };
  }
  deactivate() {
    this.onDeactivate && this.onDeactivate();
  }
  createSimulation(n, e) {
    if (this.options.createSimulation)
      return this.options.createSimulation(Jt, n, e);
    {
      const r = ot(e).id((i) => i.id);
      return ct(n).force("edge", r.distance(100)).force("charge", ht()).force("collide", it(50).strength(0.2)).force("center", et().strength(0.05)).alphaMin(1e-3);
    }
  }
  buildNodeLayouts(n, e, r) {
    const i = Object.keys(n).filter((a) => !(a in e.value));
    for (const a of i)
      e.value[a] = V({}, r);
    const o = this.forceNodeLayouts(e.value), l = Object.fromEntries(o.map((a) => [a.id, a]));
    return { nodeLayouts: o, nodeLayoutMap: l };
  }
  forceNodeLayouts(n) {
    return Object.entries(n).map(([e, r]) => r.fixed ? rn(V({ id: e }, r), { fx: r.x, fy: r.y }) : V({ id: e }, r));
  }
  forceLayoutEdges(n, e) {
    return Object.values(n).filter((r) => r.source in e && r.target in e).map((r) => ({
      source: r.source,
      target: r.target
    }));
  }
  getNodeLayout(n, e) {
    const r = gt(n.value, e);
    return r.value || (r.value = { x: 0, y: 0 }), r;
  }
}
export {
  an as ForceLayout
};
